### ISSUE TITLE
Unexpected Exception / Missing Exception when using label expression in the MATCH clause

### ISSUE THREAD

**Issue by joyemang33** [neo4j/neo4j#13236]
- Neo4j version: 5.8.0-enterprise
- Operating system: Ubuntu 20.04
- API/Driver: Cypher
- **Steps to reproduce**

Hello! While testing Neo4j, I met inconsistent behaviors when executing equivalent queries

The first query will result in an exception, which is with a different message from the exception reported in https://github.com/neo4j/neo4j/issues/13234
```cypher
MATCH (n2)-[:(!T6&!(T0&(T2|T6)))]-(n3:(!L2|!L7))-[]-(n4) RETURN *
```
![1687878941507](https://github.com/neo4j/neo4j/assets/37923722/c185e39e-13a9-47bf-a7de-41ccf2005fc6)

However, when executing the second query, it is OK to return results, and the both queries are equivalent

```cypher
MATCH (n3:(!L2|(!L7)))-[]-(n4), (n2)-[:(!T6&!(T0&(T2|T6)))]-(n3) RETURN *
```
![1687879016914](https://github.com/neo4j/neo4j/assets/37923722/cf1936de-71a8-4472-913d-a7220924388c)

Would it is possible for you to further confirm and investigate this inconsistent issue? It would be highly appreciated.

The graph is created by the following Cypher statements.

[create-2.log](https://github.com/neo4j/neo4j/files/11884366/create-2.log)

Best regards,
Joye

**Comment by Unfiled6748** [comment]
Hi joyemang33, I don't work for Neo4J, but in their docs, they mention the boolean operators are NOT, AND, OR, and XOR (see here: https://neo4j.com/docs/cypher-manual/current/syntax/operators/#query-operators-boolean). I ran the following:
```
# Query 1
MATCH (n2)-[r]-(n3)-[]-(n4) 
WHERE NOT type(r) = 'T6' AND NOT (type(r) = 'T0' AND (type(r) = 'T2' OR type(r) = 'T6')) 
AND NOT (n3:L2 OR n3:L7)
RETURN *;
```

```
# Query 2
MATCH (n3)-[]-(n4), (n2)-[r]-(n3) 
WHERE NOT type(r) = 'T6' AND NOT (type(r) = 'T0' AND (type(r) = 'T2' OR type(r) = 'T6')) 
AND NOT (n3:L2 OR n3:L7)
RETURN *;
```

and I got 685 rows returned for both of them.

Neo4j version: 5.8.0-community
Operating system: Apple M1
API/Driver Cypher

**Comment by joyemang33** [comment]
Hello @Unfiled6748 !
Many thanks for your help, and I think your query is correct.
But I think this issue is related to a bug in label expression (another way to write this query) associated with the relationship.
(see here: https://neo4j.com/docs/cypher-manual/current/patterns/reference/#label-expressions).

Best regards,
Joye

**Comment by Unfiled6748** [comment]
Thank you @joyemang33 for pointing to the documentation about label expressions. I am new to Neo4J , and I was not aware of this feature. I'll take some time to read about it!

**Comment by pontusmelke** [comment]
Hello @joyemang33, thanks for reporting!

It looks like the issue has already been resolved, can you please try again using Neo4j 5.9.0?

Best regards,
Pontus
